src/JSON.jl

```diff
@@ -68,116 +131,88 @@ function end_object(io::IO, state::State{NOINDENT}, is_dict::Bool)
 end
 
 function print_escaped(io::IO, s::AbstractString)
-    for c in s
-        c <= '\x7f' ? (unescaped[@compat(Int(c))+1]  ? Base.print(io, c) :
-                       c == '\\'       ? Base.print(io, "\\\\") :
-                       c == '"'        ? Base.print(io, "\\\"") :
-                       8 ≤ @compat(UInt32(c)) ≤ 10 ? Base.print(io, '\\', "btn"[@compat(Int(c))-7]) :
-                       c == '\f'       ? Base.print(io, "\\f") :
-                       c == '\r'       ? Base.print(io, "\\r") :
-                                         Base.print(io, "\\u", hex(c, 4))) :
+    @inbounds for c in s
+        c <= '\x7f' ? Base.write(io, escaped[@compat UInt8(c) + 0x01]) :
                       Base.print(io, c) #JSON is UTF8 encoded
     end
 end
 
-function _print(io::IO, state::State, s::AbstractString)
-    Base.print(io, '"')
-    JSON.print_escaped(io, s)
-    Base.print(io, '"')
-end
-
-if isdefined(Base, :Dates)
-    function _print(io::IO, state::State, s::Base.Dates.TimeType)
-        _print(io, state, string(s))
+function print_escaped(io::IO, s::Compat.UTF8String)
+    @inbounds for c in s.data
+        Base.write(io, escaped[c + 0x01])
     end
 end
 
-function _print(io::IO, state::State, s::Symbol)
-    _print(io, state, string(s))
+function _writejson(io::IO, state::State, s::AbstractString)
+    Base.print(io, '"')
+    JSON.print_escaped(io, s)
+    Base.print(io, '"')
 end
 
-@compat function _print(io::IO, state::State, s::Union{Integer, AbstractFloat})
+@compat function _writejson(io::IO, state::State, s::Union{Integer, AbstractFloat})
     if isnan(s) || isinf(s)
         Base.print(io, "null")
     else
         Base.print(io, s)
     end
 end
 
-@compat function _print(io::IO, state::State, n::Void)
+@compat function _writejson(io::IO, state::State, n::Void)
     Base.print(io, "null")
 end
 
-function _print(io::IO, state::State, a::Associative)
+function _writejson(io::IO, state::State, a::Associative)
     if length(a) == 0
         Base.print(io, "{}")
         return
     end
     start_object(io, state, true)
     first = true
-    for (key, value) in a
+    for key in keys(a)
```

Comment:
Benchmarking shows that the tuple is cheaper (though it allocates three times the memory), but the cost of hashing is peanuts compared to the cost of IO in this loop, so the difference isn't detectable. I switched this because for `AssociativeWrapper`, the "hash" is just a fast `getfield` call; forming the tuple still has allocation issues. After #162, it will be possible to define separate serialization for `AssociativeWrapper` and `Dict`, so I switched back in that PR.